src/libappimage/libappimage.cpp

```diff
@@ -32,29 +32,39 @@ using namespace appimage::utils;
 
 namespace bf = boost::filesystem;
 
+/**
+ * We cannot allow any exception to scape from C++ to C. This will wrap a given code block into a try-catch
+ * structure and will effectively catch and log all exceptions.
+ *
+ * @param src
+ */
+#define CATCH_ALL(src) { \
```

Comment:
This is not what I meant. I suggested just a macro that logs a single exception. That macro should be defined in the logger header, i.e.:

```cpp
#define LOG_EXCEPTION(e) Logger::exception(e, __FUNCTION__);
```

This could be used here, too. Of course it requires a `Logger::exception` method.

Reply:
I know, the problem was the code duplication now it's gone.

I would preffer to not "overfit" the `Logger` class API with this utility macro. The macro will also require the log lever so it would be something like:
```
#define LOG_EXCEPTION(level, e) Logger::log(level,  __FUNCTION__ + " : " + e.wath());
...

LOG_EXCEPTION(LogLevel::WARNING, e)
```
Which is not a big improvement over:
```
Logger::warning( __FUNCTION__ + " : " + e.wath())
```

Reply:
This is a matter of style. You didn't pick up my idea regarding `Logger::exception` at all. And exceptions are not _warnings_, they're _errors_.

Reply:
> And exceptions are not warnings, they're errors.

I have to disagree here, if exceptions were errors, they will be called that way. They are "exceptional" situations therefore depending on its nature they could be logged using any of the available logging levels.

Reply:
Of course exceptions are errors, they must per definition not be used for anything like "warnings" etc., anything like that is bad software architecture. And we're speaking of log levels here. `error` is the right one for any exception.

Reply:
> Of course exceptions are errors

Error handling is the most common case of an exceptional situation but not the only one. In fact, they are another mechanism to "transfer control and information". It has nothing to do with bad software architecture, on the contrary, is about using all the language capabilities.
Please read the first lines of
- http://www.cplusplus.com/doc/tutorial/exceptions/
- https://en.cppreference.com/w/cpp/language/exceptions


Even if you think on them as errors they don't have the same severity.

Reply:
Let me put it like this: Exceptions should not be used excessively for non-errors in C++, since they're slow. They shouldn't be "normal" (and thus not part of the normal program flow), they should occur very rarely (ideally never). For anything else, there's return values. Semantic information should not be transported over exceptions.

Whenever you _have_ to log an exception (with it's `what()`), it's in a `catch` block where you return an error state in the C interface or anything like that. If the exception were "normal", then why log it at all (or log its cause)? Therefore the `Logger::exception()` can be assumed to be used only in error states. And don't overly weigh the log levels: pretty much everyone ignores the words that describe them, there can be normal log messages on error level just to make sure the user receives them, too.

Reply:
This is about to become a theoretical discussion, mind if we continue it in a separated issue?

Please, let me know if there is something else that should be reviewed.

Reply:
Yes. I'd like to have the `exception()` log convenience helper.

Reply:
Is this what you want to be added to the public libappimage API ?
```
class Logger {
...
static void exception(const std::runtime_error& error, const char* fname);
...
}

#define LOG_EXCEPTION(e) Logger::exception(e, __FUNCTION__);
```

Let me highlight a few things:
- Preprocessor macros don't respect c++ namespaces, therefore this will pollute the global environment of liappimage users. So it will have to be called `LIBAPPIMAGE_LOG_EXCEPTION` which is no longer a nice and short name.
- the `exception` method has a different signature than the rest of the utility functions which breaks the style. 
- this will not provide any significant improvement in the libappimage code.
- using the libappimage logger outside libappimge (by the clients) seems like a really bad idea, our purpose with libappimage should never be to provide a logging library.

Said that, please confirm your intention and I'll add the feature in question.

Reply:
There's an overload missing for other exceptions, but yes.

> Preprocessor macros don't respect c++ namespaces, therefore this will pollute the global environment of liappimage users. So it will have to be called LIBAPPIMAGE_LOG_EXCEPTION which is no longer a nice and short name.

The macro doesn't have to be placed in a "global" header.

> the exception method has a different signature than the rest of the utility functions which breaks the style.

... what? What style?

> this will not provide any significant improvement in the libappimage code.

Any other logger API has such a method. And I think it's useful.

> using the libappimage logger outside libappimge (by the clients) seems like a really bad idea, our purpose with libappimage should never be to provide a logging library.

As said, the logger API doesn't have to be public. You designed it that way, with your singleton thingy. Of course it's not necessary.

Reply:
> You designed it that way, with your singleton thingy.

I guess that we both agree on making it public, but yes it is a bad idea. I'll make it private again.

If the reason for having an `void exception(const std::runtime_error& error, const char* fname);` utility method is because every decent logger has one. This logger is not meant to be decent it's just meant to fulfill the `libappimage` logging needs. Inside libappimage there is no need for it.